### PR TITLE
Port l10n_ch_scan_bvr to 8.0

### PR DESCRIPTION
**Pull request by tafaRU**
This PR replaces https://github.com/OCA/l10n-switzerland/pull/98

### CODE REVIEW

**Comment by tafaRU** [comment]
@yvaucher could you please set it as WIP till I have finished?
Thank you in advance.

**Comment by tafaRU** [comment]
@yvaucher, I ported the module to the new API and I also fixed the test.
This WARNING: https://travis-ci.org/OCA/l10n-switzerland/jobs/84101039#L317 has raised because there is no field `bvr_number` on model `res.partner.bank`.
Same thing for the field `post_number`. See https://github.com/tafaRU/l10n-switzerland/blob/8.0-l10n_ch_scan_bvr-add/l10n_ch_scan_bvr/wizard/scan_bvr.py#L208-L213
How can I manage these lines?

Thank you in advance.

**Comment by yvaucher** [comment]
You mean `post_number` and `bvr_number`?

`post_number` has been renamed ccp since 7.0 (it was missed in 7.0 port of this module)

https://github.com/tafaRU/l10n-switzerland/blob/8.0-l10n_ch_scan_bvr-add/l10n_ch_base_bank/bank.py#L169-L177

And `bvr_number` was removed in 6.0 using the same field `post_number` thus it needs to be `ccp` well.
https://code.launchpad.net/~camptocamp/openobject-addons/6.0_l10n_ch_continuous_improvment/+merge/50128

**Comment by tafaRU** [comment]
@yvaucher, thanks a lot for your explanation!
I think now PR is ready for review, so you could change its label.

**Comment by yvaucher** [comment]
@tafaRU Thanks for the work I'll have a deeper look at it on Friday or next week

**Comment by open-net-sarl** [comment]
Hi

May I suggest that you display the amount in the wizard's popup? We had add it in our version, as a request from a few of our customers.
That being said, may I suggest a code correction? 
The wizard's purpose is to create a supplier invoice. Even though you set the right type, you should consider using the partner's supplier payment conditions (i.e. the field "property_supplier_payment_term", instead of "property_payment_term").
Furthermore, I think that the context should reflect the type 'in_invoice' instead of 'out_invoice' while exiting the function "_create_direct_invoice(...)"

File: l10n_ch_scan_bvr/README.rst [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+8.0%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
+
+
+Credits
+=======
+
+Contributors
+------------
+
+* Alex Comba <alex.comba@agielbg.com>
```
Comment (yvaucher) [inline l10n_ch_scan_bvr/README.rst]: Correct me if I'm wrong but some more people deserve to be in there

Comment (yvaucher) [inline l10n_ch_scan_bvr/README.rst]: https://github.com/OCA/l10n-switzerland/commits/7.0/l10n_ch_scan_bvr

File: l10n_ch_scan_bvr/wizard/scan_bvr.py [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+                {'ccp': data['bvr_struct']['beneficiaire']})
+        else:
+            partner_bank.write(
+                {'bvr_adherent_num': data['bvr_struct']['bvrnumber'],
+                 'ccp': data['bvr_struct']['beneficiaire']})
+        date_due = time.strftime('%Y-%m-%d')
+        # We will now compute the due date and fixe the payment term
+        payment_term_id = (account_info.partner_id.property_payment_term and
+                           account_info.partner_id.property_payment_term.id or
+                           False)
```
Comment (yvaucher) [inline l10n_ch_scan_bvr/wizard/scan_bvr.py]: No need for the `and` and `or` as `account_info.partner_id.property_payment_term.id` will be either an id or False

File: l10n_ch_scan_bvr/wizard/scan_bvr.py [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+        # Account Modification
+        partner_bank = partner_bank_model.browse(data['bank_account'])
+        if data['bvr_struct']['domain'] == 'name':
+            partner_bank.write(
+                {'ccp': data['bvr_struct']['beneficiaire']})
+        else:
+            partner_bank.write(
+                {'bvr_adherent_num': data['bvr_struct']['bvrnumber'],
+                 'ccp': data['bvr_struct']['beneficiaire']})
+        date_due = time.strftime('%Y-%m-%d')
```
Comment (yvaucher) [inline l10n_ch_scan_bvr/wizard/scan_bvr.py]: Please use `fields.Date` helper `today` to use the right string format from constants

https://www.odoo.com/documentation/8.0/reference/orm.html#openerp.fields.Date

File: l10n_ch_scan_bvr/wizard/scan_bvr.py [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+        # We will now compute the due date and fixe the payment term
+        payment_term_id = (account_info.partner_id.property_payment_term and
+                           account_info.partner_id.property_payment_term.id or
+                           False)
+        if payment_term_id:
+            # We Calculate due_date
+            res = invoice_model.onchange_payment_term_date_invoice(
+                payment_term_id,
+                time.strftime('%Y-%m-%d')
```
Comment (yvaucher) [inline l10n_ch_scan_bvr/wizard/scan_bvr.py]: You could compute today once and reuse the variable

File: l10n_ch_scan_bvr/wizard/scan_bvr.py [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+            )
+            date_due = res['value']['date_due']
+
+        invoice_vals = {
+            'name': time.strftime('%Y-%m-%d'),
```
Comment (yvaucher) [inline l10n_ch_scan_bvr/wizard/scan_bvr.py]: Same here for today date

File: l10n_ch_scan_bvr/wizard/scan_bvr.py [hunk 6]
```diff
@@ -0,0 +1,431 @@
...
+            'partner_id': account_info.partner_id.id,
+            'account_id': account_info.partner_id.property_account_payable.id,
+            'date_due': date_due,
+            'date_invoice': time.strftime('%Y-%m-%d'),
```
Comment (yvaucher) [inline l10n_ch_scan_bvr/wizard/scan_bvr.py]: And here again for today date

File: l10n_ch_scan_bvr/README.rst (hunk 1 above)
Comment (tafaRU) [inline l10n_ch_scan_bvr/README.rst]: @yvaucher, I didn't know during the module migration I had to watch the commit history in the previous version. 
Anyway, I have no problem to add them as contributors :wink:

File: l10n_ch_scan_bvr/wizard/scan_bvr.py [hunk 7]
```diff
@@ -0,0 +1,431 @@
...
+            account_unformated[0:2],
+            str(int(account_unformated[2:acc_len - 1])),
+            account_unformated[acc_len - 1:acc_len]
+        )
+        return account_formated
+
+    def _get_bvr_structurated(self, bvr_string):
+        if bvr_string is not False:
+            # We will get the 2 frist digit of the BVr string in order
+            # to now the BVR type of this account
```
Comment (yvaucher) [inline l10n_ch_scan_bvr/wizard/scan_bvr.py:270]: s/now/know

File: l10n_ch_scan_bvr/wizard/scan_bvr.py [hunk 8]
```diff
@@ -0,0 +1,431 @@
...
+        account_formated = "%s-%s-%s" % (
+            account_unformated[0:2],
+            str(int(account_unformated[2:acc_len - 1])),
+            account_unformated[acc_len - 1:acc_len]
+        )
+        return account_formated
+
+    def _get_bvr_structurated(self, bvr_string):
+        if bvr_string is not False:
+            # We will get the 2 frist digit of the BVr string in order
```
Comment (yvaucher) [inline l10n_ch_scan_bvr/wizard/scan_bvr.py]: s/frist/first
s/digit/digits
s/BVr/BVR

File: l10n_ch_scan_bvr/wizard/scan_bvr.py [hunk 9]
```diff
@@ -0,0 +1,431 @@
...
+            # to now the BVR type of this account
+            bvr_type = bvr_string[0:2]
+            if bvr_type == '01' and len(bvr_string) == 42:
+                # This BVR is the type of BVR in CHF
+                # WE will call the function and Call
+                bvr_struct = self._construct_bvr_postal_in_chf(bvr_string)
+                # We will test if the BVR have an Adherent Number if not we
```
Comment (yvaucher) [inline l10n_ch_scan_bvr/wizard/scan_bvr.py]: s/have/has

File: l10n_ch_scan_bvr/wizard/scan_bvr.py [hunk 10]
```diff
@@ -0,0 +1,431 @@
...
+                    bvr_struct['domain'] = 'name'
+                else:
+                    bvr_struct['domain'] = 'bvr_adherent_num'
+                # We will set the currency , in this case it's allways CHF
+                bvr_struct['currency'] = 'CHF'
+            elif bvr_type == '01':
+                # This BVR is the type of BVR in CHF
+                # We will call the function and Call
+                bvr_struct = self._construct_bvr_in_chf(bvr_string)
+                # We will test if the BVR have an Adherent Number if not
```
Comment (yvaucher) [inline l10n_ch_scan_bvr/wizard/scan_bvr.py]: s/have/has

File: l10n_ch_scan_bvr/wizard/scan_bvr.py [hunk 11]
```diff
@@ -0,0 +1,431 @@
...
+            # to now the BVR type of this account
+            bvr_type = bvr_string[0:2]
+            if bvr_type == '01' and len(bvr_string) == 42:
+                # This BVR is the type of BVR in CHF
+                # WE will call the function and Call
+                bvr_struct = self._construct_bvr_postal_in_chf(bvr_string)
+                # We will test if the BVR have an Adherent Number if not we
+                # will make the search of the account base on
+                # his name non base on the BVR adherent number
+                if (bvr_struct['bvrnumber'] == '000000'):
```
Comment (yvaucher) [inline l10n_ch_scan_bvr/wizard/scan_bvr.py]: Can we extract this from this big `if..elif`? This is copy/pasted 7 times

File: l10n_ch_scan_bvr/wizard/scan_bvr.py [hunk 12]
```diff
@@ -0,0 +1,431 @@
...
+        # <010001000060190> 052550152684006+ 43435>
+        #
+        # Explode and check  the BVR Number and structurate it
+        #
+        data = {}
+        data['bvr_struct'] = self._get_bvr_structurated(self.bvr_string)
+        # We will now search the account linked with this BVR
+        if data['bvr_struct']['domain'] == 'name':
+            domain = [('acc_number', '=', data['bvr_struct']['beneficiaire'])]
+            partners_bank = self.pool.get('res.partner.bank').search(domain)
```
Comment (yvaucher) [inline l10n_ch_scan_bvr/wizard/scan_bvr.py]: s/self.pool.get/self.env[]

Comment (yvaucher) [inline l10n_ch_scan_bvr/wizard/scan_bvr.py]: Or simply remove this line and do the search outide the `if....else`

**Comment by yvaucher** [comment]
I got this on Runbot when trying to create an invoice with the wizard:

```
Traceback (most recent call last):
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3116639-169-364eeb/openerp/http.py", line 537, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3116639-169-364eeb/openerp/http.py", line 588, in dispatch
    return self._json_response(result)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3116639-169-364eeb/openerp/http.py", line 526, in _json_response
    body = simplejson.dumps(response)
  File "/srv/openerp/instances/openerp-oca-runbot/sandbox/local/lib/python2.7/site-packages/simplejson/__init__.py", line 370, in dumps
    return _default_encoder.encode(obj)
  File "/srv/openerp/instances/openerp-oca-runbot/sandbox/local/lib/python2.7/site-packages/simplejson/encoder.py", line 269, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/srv/openerp/instances/openerp-oca-runbot/sandbox/local/lib/python2.7/site-packages/simplejson/encoder.py", line 348, in iterencode
    return _iterencode(o, 0)
  File "/srv/openerp/instances/openerp-oca-runbot/sandbox/local/lib/python2.7/site-packages/simplejson/encoder.py", line 246, in default
    raise TypeError(repr(o) + " is not JSON serializable")
TypeError: account.invoice(13,) is not JSON serializable
```

**Comment by yvaucher** [comment]
@tafaRU Since `__unported__` folder was removed, you will need to rebase this.

**Comment by tafaRU** [comment]
@open-net-sarl, could you provide a PR using my branch as target?

**Comment by tafaRU** [comment]
Thanks for the review, @yvaucher. Your comments have been addressed, please check it again.

**Comment by tafaRU** [comment]
runbot failing, but seems not related with this PR.

**Comment by yvaucher** [comment]
@tafaRU Tested, and it works :+1:
I tested use of wizard with a new partner + use of wizard with existing partner.

**Comment by guewen** [comment]
:+1:

**Comment by mdietrichc2c** [comment]
:+1:

**Comment by tafaRU** [comment]
@yvaucher, nice, thanks for the merge!

**Comment by yvaucher** [comment]
@tafaRU Thanks to you

**Pull request merged**